### ISSUE TITLE
[Bug]: Svelte component doesn't accept content changes

### ISSUE THREAD

**Issue by benbucksch** [ueberdosis/tiptap#4918]
### Which packages did you experience the bug in?

svelte

### What Tiptap version are you using?

2.2.2

### What’s the bug you are facing?

There is no official Svelte component for TipTap.

The documentation has only [sample code for a Svelte component](https://tiptap.dev/docs/editor/installation/svelte). However, that sample code has the HTML content hardcoded. Obviously, you'd make this a component property `html`. You'd also need to get the content that the user wrote out of it - that's the whole point of the editor. That's the bug here: This should be part of the sample code.

If we try to add that functionality, that gives us:
```
/** The editor content that the user writes.
 * in/out */
export let html: string;
...
      content: html,
      onTransaction: () => {
        // force re-render so `editor.isActive` works as expected
        editor = editor
      },
      onUpdate: () => {
        html = editor.getHTML();
      },
```

However, this code has a bug: If the `html` property is changed by the parent component, then the changes do not apply. This is because the code `content: html,` runs `onMount()` only.

Important, this is needed only only when parents wants to specifically change the content, but also when we simply change records / data. However, when only the property value changes, Svelte re-uses existing components and does not re-create them, so `onMount` doesn't run, so `html` is not sent to the editor, so the changes are ignored.

Straight-forward fixes in standard Svelte way, like:
```
$: editor && editor.commands.setContent(html);
```
do *not* work. When I do this, I can no longer type spaces, or rather they are stripped after I type them (because `onUpdate` sets `html`, which then triggers `setContent(html)` again, which for some unknown reason strips the spaces).

Given given how Svelte works, this is the most basic usage of the TipTap editor in Svelte, and *every* developer will need to solve this. This should be part of the sample code Svelte component. Or, better yet, there should be a standard TipTap Svelte component.

### What browser are you using?

Chrome

### Code example

https://tiptap.dev/docs/editor/installation/svelte

### What did you expect to happen?

When `html` changes (in `<TipTap bind:html={html} />`), the editor content reflects that new data.

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [ ] Yes, I’m a sponsor. 💖

**Comment by mhanbali** [comment]
Hey Ben,

I ran into something similar last night and created a solution for my specific need. I know it's not a complete solution, but in case it fits anyone else's problem.

Summary: I created two exported functions - one to insert text, used by a button. Another to bind the editor contents to another text field to mirror input. The latter I don't need/use, but just wrote real quick as a test.

In your TipTap editor component, create/export two functions outside of onMount:


```
// this will be called by the parent component in a button
export function insertText(text: string) {
	editor?.commands.insertContent(text);
}

// this will be used by a textarea in the parent with on:input
// note: this will not account for formatting such as new lines
export function changeContent(text: string) {
	editor?.commands.setContent(text);
}
```

Next, in your parent component, likely where you are calling the `<Editor />` (or whatever you name it) component, create two variables:

`$: changeText = '';`
this is to bind/update the text

`let component: Editor;`
this is to bind to the actual component. Replace "Editor" with whatever your import name is. My import of the editor looks like this:
`import Editor from '$lib/components/Editor.svelte';`

If you aren't using TypeScript, you can probably just do `let component;`

Now, bind your component to use the exported functions.
`<Editor bind:this={component} />`

Create a textarea/input. I'll use it to get the text I want to insert in this example:
```
<textarea
	rows="10"
	bind:value={changeText}
></textarea>
```

In a button you can call insertText:
`<button on:click={() => component.insertText(changeText)}>Insert Text</button>`

And that's it.

If you want to bind the textarea to the Editor component, you can add this to your `<textarea>`
`on:input={() => component.insertText(changeText)}`

Note: as said before, it won't keep formatting like new lines the way it's currently written.

**Comment by benbucksch** [comment]
> it won't keep formatting like new lines the way it's currently written.

Right. My first solution had the exact same problem. I need the formatting and new lines.

Also, I need to bind to a store (e.g. `$html`) that contains the original text and shall contain the changes. I.e. it the editor content should always match that variable, both in and out. I fear that this would create an infinite loop: The editor writes to `$html`, which updates the editor with the new value etc.. I guess we could break that loop by saving the last `html` string value in a component variable (untested code, for illustration purposes only):
```
var currentHTML: string = null;
// this will be called by the parent component in a button
export function insertText(html: string) {
  if (html == currentHTML) {
    return; // breaks infinite update loop
  }
  currentHTML = html;
  editor?.commands.insertContent(html);
}

// Attention: This will not save formatting nor new lines
export function changeContent(html: string) {
  currentHTML = html;
  editor?.commands.setContent(html);
}
```

**Comment by benbucksch** [comment]
> it won't keep formatting like new lines the way it's currently written.

Right. My first solution had the exact same problem, see initial description: "`setContent(html)` again, which for some unknown reason strips the spaces."

I need the formatting and new lines, so this solution didn't work. Obviously, an HTML editor where I cannot get the content out nor not the formatting, is kind-of pointless. That's why I filed this bug here.

**Comment by mhanbali** [comment]
Try updating the function to the below. It looks to keep the new lines:

```
export function changeContent(text: string) {
  editor?.commands.clearContent();
  editor?.commands.insertContent(text);
}
```

**Comment by benbucksch** [comment]
After years of struggling with this, I finally have something that works. Here's the Svelte component:

```
<div bind:this={rootEl} class="html-editor" />

<script lang="ts">
  import { Editor } from '@tiptap/core';
  import StarterKit from '@tiptap/starter-kit';
  import { onMount, onDestroy } from 'svelte';

  /** in/out */
  export let html: string;
  /** out only */
  export let editor: Editor;

  let rootEl: HTMLDivElement;
  let lastHTML: string = null;

  onMount(onLoad);

  function onLoad() {
    createEditor();
  }

  function createEditor() {
    editor = new Editor({
      element: rootEl,
      extensions: [
        StarterKit,
      ],
      content: html,
      onTransaction: () => {
        // force re-render so `editor.isActive` works as expected
        editor = editor;
      },
      onUpdate: ({ editor }) => {
        html = lastHTML = editor.getHTML();
      },
    });
    lastHTML = html;
  }

  export function forceReload() {
    if (editor) {
      editor.destroy();
    }
    createEditor();
  }

  // html changed by caller, not editor
  $: html != lastHTML && forceReload();

  onDestroy(() => {
    if (editor) {
      editor.destroy();
    }
  });
</script>

<style>
  .html-editor {
    height: 100%;
  }
  .html-editor :global(.ProseMirror) {
    height: 100%;
  }
  .html-editor :global(.ProseMirror:focus-visible) {
    outline: none;
  }

  /** Undo the default margin of first/last <p> */
  .html-editor :global(.tiptap) {
    margin-top: -1em;
    margin-bottom: -1em;
  }
</style>
```

**Comment by kotekpsotek** [comment]
https://github.com/ueberdosis/tiptap/issues/4918#issuecomment-3051042127 Actually, it's the bulletproof solution. Works everytime. Idk why [svelte-tiptap](https://www.npmjs.com/package/svelte-tiptap) hasn't got implemented it yet
